### PR TITLE
feat: Add OIDC authentication

### LINKED ISSUE #695: AuthN / OIDC
These are the items we identify as needed in the Enterprise MVP as wither paid feature or additional Community features. 

-----

- [x] https://github.com/coder/coder/issues/820
- [x] #789
- [x] https://github.com/coder/coder/issues/790
- [x] https://github.com/coder/coder/issues/792
- [ ] https://github.com/coder/coder/issues/696

### PR DESCRIPTION
**Pull request by kylecarbs**
This adds OIDC authentication to the product!

The email and username cannot be used as unique, so we'll have to add an identifier to match a user. See: https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability. https://github.com/coder/coder/issues/3322 has been created to fix this.

Closes #695.

https://user-images.githubusercontent.com/7122116/182042186-45b69b84-a563-46ff-8155-4fd13c409671.mp4

Most of the diff is extracting the username validation into its package; the OIDC exchange is just a few hundred lines. This was manually tested with Google, but considering the extent of tests, it'll hopefully work with others!

### CODE REVIEW

**Comment by ntimo** [comment]
This PR made my day ♥️ Can‘t wait to use OIDC 👍

File: coderd/database/dump/main.go [hunk 1]
```diff
@@ -31,6 +31,11 @@ func main() {
 	}
 
 	cmd := exec.Command(
+		"docker",
+		"run",
+		"--rm",
+		"--network=host",
+		"postgres:13",
```
Comment (kylecarbs) [inline coderd/database/dump/main.go:38]: This was a bit out of scope, but makes everything use Docker, which makes changing our versions simpler!

File: coderd/httpapi/httpapi_test.go [hunk 2]
```diff
@@ -81,71 +81,6 @@ func TestRead(t *testing.T) {
 	})
 }
 
-func TestReadUsername(t *testing.T) {
```
Comment (kylecarbs) [inline coderd/httpapi/httpapi_test.go:84]: These have been moved to the `username` package!

File: docs/install/auth.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Authentication
+
+By default, Coder is accessible via password authentication. Alternatively, you
+can configure Coder to enable logging in through GitHub OAuth or OpenID Connect.
+See below for configuration steps.
```
Comment (ammario) [inline docs/install/auth.md:5]:
```suggestion
By default, Coder is accessible via password authentication.

The following steps explain how to set up GitHub OAuth or OpenID Connect.
```

File: docs/install/auth.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+CODER_OAUTH2_GITHUB_ALLOWED_ORGS="your-org"
+CODER_OAUTH2_GITHUB_CLIENT_ID="8d1...e05"
+CODER_OAUTH2_GITHUB_CLIENT_SECRET="57ebc9...02c24c"
+```
+
+Once complete, run `sudo service coder restart` to reboot Coder.
+
+## OpenID Connect with Google
+
+> Any OpenID connect provider can be used with Coder (e.g. Okta, Azure Active Directory, GitLab, Auth0). This example with Google should be used as guidance.
```
Comment (ammario) [inline docs/install/auth.md:43]:
```suggestion
> We describe how to set up the most popular OIDC provider, Google, but any (Okta, Azure Active Directory, GitLab, Auth0, etc.) may be used.
```

**Review by ammario — CHANGES_REQUESTED** [review]

File: coderd/userauth.go [hunk 5]
```diff
@@ -205,3 +209,137 @@ func (api *API) userOAuth2Github(rw http.ResponseWriter, r *http.Request) {
...
+		if !strings.HasSuffix(claims.Email, api.OIDCConfig.EmailDomain) {
+			httpapi.Write(rw, http.StatusForbidden, codersdk.Response{
+				Message: fmt.Sprintf("Your email %q is not a part of the %q domain!", claims.Email, api.OIDCConfig.EmailDomain),
+			})
+			return
+		}
+	}
+
+	var user database.User
+	user, err = api.Database.GetUserByEmailOrUsername(r.Context(), database.GetUserByEmailOrUsernameParams{
```
Comment (sreya) [inline coderd/userauth.go:290]: This is pretty unlikely to happen but if an existing, unrelated user (say with `build-in` auth) already has this email address, then it's possible for a new OIDC user to impersonate them by providing the same email address.

File: coderd/username/username.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package username
```
Comment (sreya) [inline coderd/username/username.go:1]: I like the idea of extracting validation to a separate package but what sort of precedent do we want to be setting if we have other fields that may be suitable for relocation to a new package? Are we expecting each new field to create its own pkg, to refactor into a common `validate` pkg or do you feel like `username` is sufficiently unique to warrant its own pkg and most everything else should be contained to `coderd`?

File: coderd/userauth.go (hunk 5 above)
Comment (kylecarbs) [reply coderd/userauth.go:290]: Yup. The issue I attached expands on that a bit. We'll have to make the user-identification with account links better.

File: cli/server.go [hunk 7]
```diff
@@ -282,6 +289,31 @@ func server() *cobra.Command {
 				}
 			}
 
+			if oidcClientSecret != "" {
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply cli/server.go:292]: Yeah, I'll improve this to make it more obvious the combination of flags you need.

File: coderd/username/username.go (hunk 6 above)
Comment (kylecarbs) [reply coderd/username/username.go:1]: I feel like this isn't just a case of validation. I extracted this primarily due to the generation that we now need to do with OIDC.

The alternative was exposing `httpapi.ValidateUsername` or something of the sort, which I'm also mostly fine with. Keep in mind we'd have to have a separate function of `httpapi.UsernameFromString` or something like it, but I'm also fine with that. I'll think on it, but if you think that sounds cleaner I'll make it happen!

File: coderd/userauth.go (hunk 5 above)
Comment (sreya) [reply coderd/userauth.go:290]: Do we want to add a column to the users table that indicates the sort of authentication they used? Probably not necessary to do in this PR

File: coderd/username/username.go (hunk 6 above)
Comment (sreya) [reply coderd/username/username.go:1]: Nah I was just bringing it up to get an idea of how we want to structure packages

**Review by sreya — APPROVED** [review]

File: coderd/userauth.go (hunk 5 above)
Comment (kylecarbs) [reply coderd/userauth.go:290]: We will in a future PR, this one just gets it out the door

File: coderd/username/username.go (hunk 6 above)
Comment (kylecarbs) [reply coderd/username/username.go:1]: I actually removed the username package. It felt too much like floaty scope with little benefit. That code is in `httpapi` now, which feels better anyways.

Good comment :+1:

**Review by ammario — APPROVED** [review]

**Pull request merged**